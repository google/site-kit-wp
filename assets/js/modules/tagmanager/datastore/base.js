/**
 * `modules/tagmanager` base data store
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Modules from 'googlesitekit-modules';
import { MODULES_TAGMANAGER } from './constants';
import { MODULE_SLUG_TAGMANAGER } from '@/js/modules/tagmanager/constants';
import { submitChanges, validateCanSubmitChanges } from './settings';

let baseModuleStore = Modules.createModuleStore( MODULE_SLUG_TAGMANAGER, {
	ownedSettingsSlugs: [
		'accountID',
		'ampContainerID',
		'containerID',
		'internalAMPContainerID',
		'internalContainerID',
	],
	storeName: MODULES_TAGMANAGER,
	settingSlugs: [
		'accountID',
		'ampContainerID',
		'containerID',
		'internalContainerID',
		'internalAMPContainerID',
		'useSnippet',
		'ownerID',
	],
	submitChanges,
	validateCanSubmitChanges,
} );

// Rename generated pieces to adhere to our convention.
baseModuleStore = ( ( { actions, selectors, ...store } ) => {
	// eslint-disable-next-line sitekit/acronym-case
	const { setAmpContainerID, ...restActions } = actions;
	// eslint-disable-next-line sitekit/acronym-case
	const { getAmpContainerID, ...restSelectors } = selectors;

	return {
		...store,
		actions: {
			...restActions,
			// eslint-disable-next-line sitekit/acronym-case
			setAMPContainerID: setAmpContainerID,
		},
		selectors: {
			...restSelectors,
			// eslint-disable-next-line sitekit/acronym-case
			getAMPContainerID: getAmpContainerID,
		},
	};
} )( baseModuleStore );

export default baseModuleStore;
