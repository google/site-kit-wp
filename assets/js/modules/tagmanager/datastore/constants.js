/**
 * `modules/tagmanager` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_TAGMANAGER = 'modules/tagmanager';

// A special Account ID value used for the "Set up a new account" option.
export const ACCOUNT_CREATE = 'account_create';
// A special Container ID value used for the "Set up a new container" option.
export const CONTAINER_CREATE = 'container_create';
// Usage context for web containers.
export const CONTEXT_WEB = 'web';
// Usage context for AMP containers.
export const CONTEXT_AMP = 'amp';
// Form ID for the module setup form.
export const FORM_SETUP = 'tagmanagerSetup';
// OAuth scope needed for viewing containers.
export const READ_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';
// OAuth scope needed for creating containers.
export const EDIT_SCOPE =
	'https://www.googleapis.com/auth/tagmanager.edit.containers';
// A special setupMode value for when completing setup with Analytics.
export const SETUP_MODE_WITH_ANALYTICS = 'SETUP_WITH_ANALYTICS';
