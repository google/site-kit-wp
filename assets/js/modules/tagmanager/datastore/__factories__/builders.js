/**
 * Tag Manager object builders.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { build, sequence, oneOf, perBuild } from '@jackfranklin/test-data-bot';
import faker from 'faker';

/**
 * Internal dependencies
 */
import { CONTEXT_WEB, CONTEXT_AMP } from '../constants';

/**
 * Creates an account object in the format returned by the list endpoint.
 *
 * @since 1.11.0
 * @see {@link https://developers.google.com/tag-manager/api/v2/reference/accounts/list}
 * @private
 *
 * @param {Object} [args]           Optional arguments to the builder.
 * @param {Object} [args.overrides] Optional fields overrides to take precedence over the default generated values.
 * @param {Object} [args.map]       Optional function for transforming the generated object.
 * @return {Object} Resulting generated account object.
 */
export const accountBuilder = build( 'Tag Manager Account', {
	fields: {
		path: 'accounts/{accountId}',
		accountId: sequence( ( num ) => `${ 100 + num }` ), // eslint-disable-line sitekit/acronym-case
		name: perBuild( () => faker.lorem.words() ),
	},
	postBuild: ( account ) => {
		const { accountId } = account; // eslint-disable-line sitekit/acronym-case
		return {
			...account,
			path: `accounts/${ accountId }`, // eslint-disable-line sitekit/acronym-case
		};
	},
} );

/**
 * Creates a container object in the format returned by the list endpoint.
 *
 * @since 1.11.0
 * @see {@link https://developers.google.com/tag-manager/api/v2/reference/accounts/containers/list}
 * @private
 *
 * @param {Object} [args]           Optional arguments to the builder.
 * @param {Object} [args.overrides] Optional fields overrides to take precedence over the default generated values.
 * @param {Object} [args.map]       Optional function for transforming the generated object.
 * @return {Object} Resulting generated container object.
 */
export const containerBuilder = build( 'Tag Manager Container', {
	fields: {
		path: 'accounts/{accountId}/containers/{containerId}',
		accountId: perBuild( () => faker.datatype.number().toString() ), // eslint-disable-line sitekit/acronym-case
		containerId: sequence( ( num ) => `${ 200 + num }` ), // eslint-disable-line sitekit/acronym-case
		name: perBuild( () => faker.lorem.words() ),
		// eslint-disable-next-line sitekit/acronym-case
		publicId: perBuild( () => {
			const char = faker.random.alphaNumeric;
			return `GTM-FAKE${ char() }${ char() }${ char() }`.toUpperCase();
		} ),
		usageContext: [ oneOf( CONTEXT_WEB, CONTEXT_AMP ) ],
		fingerprint: Date.now().toString(),
		// eslint-disable-next-line sitekit/acronym-case
		tagManagerUrl:
			'https://tagmanager.google.com/#/container/accounts/{accountId}/containers/{containerId}/workspaces?apiLink=container',
	},
	postBuild: ( container ) => {
		const { accountId, containerId } = container; // eslint-disable-line sitekit/acronym-case

		return {
			...container,
			// eslint-disable-next-line sitekit/acronym-case
			path: `accounts/${ accountId }/containers/${ containerId }`,
			// eslint-disable-next-line sitekit/acronym-case
			tagManagerUrl: `https://tagmanager.google.com/#/container/accounts/${ accountId }/containers/${ containerId }/workspaces?apiLink=container`,
		};
	},
} );

/**
 * Generates multiple containers.
 *
 * @since 1.12.0
 * @private
 *
 * @param {number} count       Number of containers to generate.
 * @param {Object} [overrides] Optional. Object of container field overrides.
 * @return {Object[]} Array of generated container objects.
 */
export function buildContainers( count, overrides ) {
	return Array.from( { length: count } ).map( () =>
		containerBuilder( { overrides } )
	);
}

/**
 * Generates an account with one or more containers.
 *
 * @since 1.11.0
 * @private
 *
 * @param {Object} [args]           Optional args for controlling the output.
 * @param {Object} [args.account]   Account field overrides.
 * @param {Object} [args.container] Container field overrides.
 * @param {number} [args.count]     Number of containers to create.
 * @return {Object} Generated account and containers { account, containers }.
 */
export function buildAccountWithContainers( {
	account: accountOverrides = {},
	container: containerOverrides = {},
	count = 1,
} = {} ) {
	const account = accountBuilder( { overrides: accountOverrides } );
	const containers = buildContainers( count, {
		...containerOverrides,
		accountId: account.accountId, // eslint-disable-line sitekit/acronym-case
	} );

	return {
		account,
		containers,
	};
}

/* eslint-disable sitekit/acronym-case */
export function defaultTagWeb( { accountId, containerId } = {} ) {
	return {
		accountId,
		blockingRuleId: null,
		blockingTriggerId: null,
		containerId,
		fingerprint: Date.now().toString(),
		firingRuleId: null,
		firingTriggerId: [ '2147479553' ],
		liveOnly: null,
		monitoringMetadataTagNameKey: null,
		name: 'Default web tag',
		notes: null,
		parentFolderId: null,
		path: null,
		paused: null,
		scheduleEndMs: null,
		scheduleStartMs: null,
		tagFiringOption: 'oncePerEvent',
		tagId: '2',
		tagManagerUrl: null,
		type: 'html',
		workspaceId: null,
		parameter: [
			{
				key: 'html',
				type: 'template',
				value: "console.log( 'Hello world!' );",
			},
			{
				key: 'supportDocumentWrite',
				type: 'boolean',
				value: 'false',
			},
		],
		monitoringMetadata: {
			key: null,
			type: 'map',
			value: null,
		},
	};
}

function defaultTagAMP( { accountId, containerId } = {} ) {
	return {
		accountId,
		blockingRuleId: null,
		blockingTriggerId: null,
		containerId,
		fingerprint: Date.now().toString(),
		firingRuleId: null,
		firingTriggerId: [ '2147479553' ],
		liveOnly: null,
		monitoringMetadataTagNameKey: null,
		name: 'Default AMP tag',
		notes: null,
		parentFolderId: null,
		path: null,
		paused: null,
		scheduleEndMs: null,
		scheduleStartMs: null,
		tagFiringOption: 'oncePerEvent',
		tagId: '2',
		tagManagerUrl: null,
		type: 'img',
		workspaceId: null,
		parameter: [
			{
				key: 'useCacheBuster',
				type: 'boolean',
				value: 'true',
			},
			{
				key: 'url',
				type: 'template',
				value: '//example.com/favicon.ico',
			},
			{
				key: 'cacheBusterQueryParam',
				type: 'template',
				value: 'gtmcb',
			},
		],
	};
}

export const liveContainerVersionBuilder = build(
	'Tag Manager Live Container Version',
	{
		fields: {
			accountId: perBuild( () => faker.datatype.number().toString() ), // Relationship
			builtInVariable: [],
			container: {
				// overrides
			},
			containerId: sequence( ( num ) => `${ 200 + num }` ),
			containerVersionId: sequence( ( num ) => `${ 0 + num }` ),
			deleted: null,
			description: null,
			name: null,
			fingerprint: Date.now().toString(),
			path: 'accounts/{accountId}/containers/{containerId}/versions/{containerVersionId}',
			tag: undefined, // required, but depends on container type.
			tagManagerUrl:
				'https://tagmanager.google.com/#/versions/accounts/{accountId}/containers/{containerId}/versions/{containerVersionId}?apiLink=version',
			variable: undefined, // absent by default.
		},
		postBuild( {
			container: containerOverrides,
			tag: tagOverride,
			...object
		} ) {
			const { accountId, containerId, containerVersionId } = object;
			const commonIDs = {
				accountId: accountId.toString(),
				containerId: containerId.toString(),
			};
			const container = containerBuilder( {
				overrides: { ...containerOverrides, ...commonIDs },
			} );
			const defaultTag =
				CONTEXT_WEB === container.usageContext[ 0 ]
					? defaultTagWeb
					: defaultTagAMP;

			return {
				...object,
				...commonIDs,
				container,
				tag: tagOverride || [ defaultTag( commonIDs ) ],
				path: `accounts/${ accountId }/containers/${ containerId }/versions/${ containerVersionId }`,
				tagManagerUrl: `https://tagmanager.google.com/#/versions/accounts/${ accountId }/containers/${ containerId }/versions/${ containerVersionId }?apiLink=version`,
			};
		},
	}
);

function googleTag( googleTagID, { accountId, containerId } = {} ) {
	return {
		accountId,
		blockingRuleId: null,
		blockingTriggerId: null,
		containerId,
		fingerprint: '1704031882618',
		firingRuleId: null,
		firingTriggerId: [ '2147479553' ],
		liveOnly: null,
		monitoringMetadataTagNameKey: null,
		name: 'Google Tag',
		notes: null,
		parentFolderId: null,
		path: null,
		paused: null,
		scheduleEndMs: null,
		scheduleStartMs: null,
		tagFiringOption: 'oncePerEvent',
		tagId: '3',
		tagManagerUrl: null,
		type: 'googtag',
		workspaceId: null,
		parameter: [
			{
				key: 'tagId',
				type: 'template',
				value: googleTagID,
			},
		],
		monitoringMetadata: {
			key: null,
			type: 'map',
			value: null,
		},
		consentSettings: {
			consentStatus: 'notSet',
		},
	};
}
function analyticsTagWeb( propertyID, { accountId, containerId } = {} ) {
	return {
		accountId,
		blockingRuleId: null,
		blockingTriggerId: null,
		containerId,
		fingerprint: Date.now().toString(),
		firingRuleId: null,
		firingTriggerId: [ '2147479553' ],
		liveOnly: null,
		monitoringMetadataTagNameKey: null,
		name: 'Google Analytics',
		notes: null,
		parentFolderId: null,
		path: null,
		paused: null,
		scheduleEndMs: null,
		scheduleStartMs: null,
		tagFiringOption: 'oncePerEvent',
		tagId: '3',
		tagManagerUrl: null,
		type: 'ua',
		workspaceId: null,
		parameter: [
			{
				key: 'overrideGaSettings',
				type: 'boolean',
				value: 'true',
			},
			{
				key: 'trackType',
				type: 'template',
				value: 'TRACK_PAGEVIEW',
			},
			{
				key: 'trackingId',
				type: 'template',
				value: propertyID,
			},
		],
		monitoringMetadata: {
			key: null,
			type: 'map',
			value: null,
		},
	};
}
function analyticsTagAMP( propertyID, { accountId, containerId } = {} ) {
	return {
		accountId,
		blockingRuleId: null,
		blockingTriggerId: null,
		containerId,
		fingerprint: Date.now().toString(),
		firingRuleId: null,
		firingTriggerId: [ '2147479553' ],
		liveOnly: null,
		monitoringMetadataTagNameKey: null,
		name: 'Google Analytics',
		notes: null,
		parentFolderId: null,
		path: null,
		paused: null,
		scheduleEndMs: null,
		scheduleStartMs: null,
		tagFiringOption: 'oncePerEvent',
		tagId: '3',
		tagManagerUrl: null,
		type: 'ua_amp',
		workspaceId: null,
		parameter: [
			{
				key: 'trackType',
				type: 'template',
				value: 'TRACK_PAGEVIEW',
			},
			{
				key: 'trackingId',
				type: 'template',
				value: propertyID,
			},
		],
	};
}
/* eslint-enable sitekit/acronym-case */

export function buildLiveContainerVersion( {
	accountID = '100',
	googleTagID,
} = {} ) {
	return liveContainerVersionBuilder( {
		overrides: {
			accountId: accountID, // eslint-disable-line sitekit/acronym-case
			container: {
				usageContext: [ CONTEXT_WEB ],
			},
		},
		map( object ) {
			if ( googleTagID ) {
				object.tag = [ googleTag( googleTagID, object ) ];
			}
			return object;
		},
	} );
}

export function buildLiveContainerVersionWeb( {
	accountID = '100',
	propertyID,
} = {} ) {
	return liveContainerVersionBuilder( {
		overrides: {
			accountId: accountID, // eslint-disable-line sitekit/acronym-case
			container: {
				usageContext: [ CONTEXT_WEB ],
			},
		},
		map( object ) {
			if ( propertyID ) {
				object.tag = [ analyticsTagWeb( propertyID, object ) ];
			}
			return object;
		},
	} );
}

export function buildLiveContainerVersionAMP( {
	accountID = '100',
	propertyID,
} = {} ) {
	return liveContainerVersionBuilder( {
		overrides: {
			accountId: accountID, // eslint-disable-line sitekit/acronym-case
			container: {
				usageContext: [ CONTEXT_AMP ],
			},
		},
		map( object ) {
			if ( propertyID ) {
				object.tag = [ analyticsTagAMP( propertyID, object ) ];
			}
			return object;
		},
	} );
}
