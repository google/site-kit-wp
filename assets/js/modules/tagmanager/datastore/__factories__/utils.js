/**
 * Tag Manager factory utils.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { MODULES_TAGMANAGER } from '../constants';
import {
	buildLiveContainerVersionWeb,
	buildLiveContainerVersionAMP,
} from './builders';

export function parseLiveContainerVersionIDs(
	{ accountId, containerId, container }, // eslint-disable-line sitekit/acronym-case
	callback
) {
	const ids = {
		accountID: accountId, // eslint-disable-line sitekit/acronym-case
		containerID: container.publicId, // eslint-disable-line sitekit/acronym-case
		ampContainerID: container.publicId, // eslint-disable-line sitekit/acronym-case
		internalContainerID: containerId, // eslint-disable-line sitekit/acronym-case
		internalAMPContainerID: containerId, // eslint-disable-line sitekit/acronym-case
	};
	if ( callback ) {
		callback( ids );
	}
	return ids;
}

export function createBuildAndReceivers( registry ) {
	return {
		buildAndReceiveWebAndAMP: ( {
			webPropertyID,
			ampPropertyID,
			accountID = '12345',
		} = {} ) => {
			const liveContainerVersionWeb = buildLiveContainerVersionWeb( {
				accountID,
				propertyID: webPropertyID,
			} );
			const liveContainerVersionAMP = buildLiveContainerVersionAMP( {
				accountID,
				propertyID: ampPropertyID,
			} );
			registry.dispatch( MODULES_TAGMANAGER ).setAccountID( accountID );
			parseLiveContainerVersionIDs(
				liveContainerVersionWeb,
				( { containerID, internalContainerID } ) => {
					registry
						.dispatch( MODULES_TAGMANAGER )
						.setContainerID( containerID );
					registry
						.dispatch( MODULES_TAGMANAGER )
						.setInternalContainerID( internalContainerID );
					registry
						.dispatch( MODULES_TAGMANAGER )
						.receiveGetLiveContainerVersion(
							liveContainerVersionWeb,
							{ accountID, internalContainerID }
						);
				}
			);
			parseLiveContainerVersionIDs(
				liveContainerVersionAMP,
				( { containerID, internalContainerID } ) => {
					registry
						.dispatch( MODULES_TAGMANAGER )
						.setAMPContainerID( containerID );
					registry
						.dispatch( MODULES_TAGMANAGER )
						.setInternalAMPContainerID( internalContainerID );
					registry
						.dispatch( MODULES_TAGMANAGER )
						.receiveGetLiveContainerVersion(
							liveContainerVersionAMP,
							{ accountID, internalContainerID }
						);
				}
			);
			return {
				accountID,
				liveContainerVersionWeb,
				liveContainerVersionAMP,
			};
		},
	};
}
