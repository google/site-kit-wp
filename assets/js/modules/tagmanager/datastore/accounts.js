/**
 * `modules/tagmanager` data store: accounts.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import {
	createRegistrySelector,
	commonActions,
	combineStores,
} from 'googlesitekit-data';
import { createValidatedAction } from '../../../googlesitekit/data/utils';
import { CORE_SITE } from '../../../googlesitekit/datastore/site/constants';
import { MODULES_TAGMANAGER, CONTAINER_CREATE } from './constants';
import { isValidAccountSelection } from '../util/validation';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { ACCOUNT_CREATE } from '../../analytics-4/datastore/constants';

// Actions
const RESET_ACCOUNTS = 'RESET_ACCOUNTS';

const fetchGetAccountsStore = createFetchStore( {
	baseName: 'getAccounts',
	controlCallback: () =>
		API.get( 'modules', 'tagmanager', 'accounts', null, {
			useCache: false,
		} ),
	reducerCallback: ( state, accounts ) => {
		return {
			...state,
			accounts,
		};
	},
} );

export const baseInitialState = {
	accounts: undefined,
};

export const baseActions = {
	/**
	 * Clears received accounts, and unsets related selections.
	 *
	 * The `getAccounts` selector will be invalidated to allow accounts to be re-fetched from the server.
	 *
	 * @since 1.12.0
	 * @private
	 */
	*resetAccounts() {
		const { dispatch } = yield commonActions.getRegistry();

		yield {
			payload: {},
			type: RESET_ACCOUNTS,
		};

		dispatch( MODULES_TAGMANAGER ).invalidateResolutionForStoreSelector(
			'getAccounts'
		);
	},

	/**
	 * Selects the given account and makes related selections.
	 *
	 * @since 1.12.0
	 * @private
	 *
	 * @param {string} accountID Tag Manager account ID to select.
	 */
	selectAccount: createValidatedAction(
		( accountID ) => {
			invariant(
				isValidAccountSelection( accountID ),
				'A valid accountID selection is required to select.'
			);
		},
		function* ( accountID ) {
			const { dispatch, select, resolveSelect } =
				yield commonActions.getRegistry();

			// Do nothing if the accountID to select is the same as the current.
			if ( accountID === select( MODULES_TAGMANAGER ).getAccountID() ) {
				return;
			}

			dispatch( MODULES_TAGMANAGER ).setSettings( {
				accountID,
				containerID: '',
				internalContainerID: '',
				ampContainerID: '',
				internalAMPContainerID: '',
			} );

			if (
				ACCOUNT_CREATE === accountID ||
				select( MODULES_TAGMANAGER ).hasExistingTag()
			) {
				return;
			}

			// Trigger cascading selections.
			const { isAMP, isSecondaryAMP } = select( CORE_SITE );
			if ( ! isAMP() || isSecondaryAMP() ) {
				const webContainers = yield commonActions.await(
					resolveSelect( MODULES_TAGMANAGER ).getWebContainers(
						accountID
					)
				);

				if ( ! webContainers.length ) {
					dispatch( MODULES_TAGMANAGER ).setContainerID(
						CONTAINER_CREATE
					);
					dispatch( MODULES_TAGMANAGER ).setInternalContainerID( '' );
				} else if ( webContainers.length === 1 ) {
					dispatch( MODULES_TAGMANAGER ).setContainerID(
						// eslint-disable-next-line sitekit/acronym-case
						webContainers[ 0 ].publicId
					);
					dispatch( MODULES_TAGMANAGER ).setInternalContainerID(
						// eslint-disable-next-line sitekit/acronym-case
						webContainers[ 0 ].containerId
					);
				}
			}

			if ( isAMP() ) {
				const ampContainers = yield commonActions.await(
					resolveSelect( MODULES_TAGMANAGER ).getAMPContainers(
						accountID
					)
				);

				if ( ! ampContainers.length ) {
					dispatch( MODULES_TAGMANAGER ).setAMPContainerID(
						CONTAINER_CREATE
					);
					dispatch( MODULES_TAGMANAGER ).setInternalAMPContainerID(
						''
					);
				} else if ( ampContainers.length === 1 ) {
					dispatch( MODULES_TAGMANAGER ).setAMPContainerID(
						// eslint-disable-next-line sitekit/acronym-case
						ampContainers[ 0 ].publicId
					);
					dispatch( MODULES_TAGMANAGER ).setInternalAMPContainerID(
						// eslint-disable-next-line sitekit/acronym-case
						ampContainers[ 0 ].containerId
					);
				}
			}
		}
	),
};

export const baseReducer = ( state, { type } ) => {
	switch ( type ) {
		case RESET_ACCOUNTS: {
			return {
				...state,
				accounts: undefined,
				settings: {
					...state.settings,
					accountID: undefined,
					ampContainerID: undefined,
					containerID: undefined,
					internalAMPContainerID: undefined,
					internalContainerID: undefined,
				},
			};
		}

		default: {
			return state;
		}
	}
};

export const baseResolvers = {
	*getAccounts() {
		const { select, dispatch } = yield commonActions.getRegistry();
		let accounts = select( MODULES_TAGMANAGER ).getAccounts();

		// Only fetch accounts if they have not been received yet.
		if ( ! accounts ) {
			( { response: accounts } =
				yield fetchGetAccountsStore.actions.fetchGetAccounts() );
		}

		if (
			accounts?.length === 1 &&
			! select( MODULES_TAGMANAGER ).getAccountID()
		) {
			dispatch( MODULES_TAGMANAGER ).selectAccount(
				// eslint-disable-next-line sitekit/acronym-case
				accounts[ 0 ].accountId
			);
		}
	},
};

export const baseSelectors = {
	/**
	 * Gets all Google Tag Manager accounts this user can access.
	 *
	 * @since 1.11.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {?Array.<Object>} An array of account objects; `undefined` if not loaded.
	 */
	getAccounts( state ) {
		const { accounts } = state;

		return accounts;
	},

	/**
	 * Checks whether accounts are currently being fetched.
	 *
	 * @since 1.11.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {boolean} Whether accounts are currently being fetched or not.
	 */
	isDoingGetAccounts: createRegistrySelector( ( select ) => () => {
		return select( MODULES_TAGMANAGER ).isFetchingGetAccounts();
	} ),
};

const store = combineStores( fetchGetAccountsStore, {
	initialState: baseInitialState,
	actions: baseActions,
	reducer: baseReducer,
	resolvers: baseResolvers,
	selectors: baseSelectors,
} );

export const {
	initialState,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
} = store;

export default store;
