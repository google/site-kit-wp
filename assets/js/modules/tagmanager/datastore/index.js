/**
 * modules/tagmanager data store
 *
 * Site Kit by Google, Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import Modules from 'googlesitekit-modules';
import { createSnapshotStore } from '../../../googlesitekit/data/create-snapshot-store';
import { STORE_NAME } from './constants';
import accounts from './accounts';
import containers from './containers';
import tags from './tags';
import settings from './settings';
import versions from './versions';
import service from './service';

export { STORE_NAME };

let baseModuleStore = Modules.createModuleStore( 'tagmanager', {
	storeName: STORE_NAME,
	settingSlugs: [
		'accountID',
		'ampContainerID',
		'containerID',
		'internalContainerID',
		'internalAMPContainerID',
		'useSnippet',
		'ownerID',
	],
} );

// Rename generated pieces to adhere to our convention.
baseModuleStore = ( ( { actions, selectors, ...store } ) => {
	const { setAmpContainerID, ...restActions } = actions;
	const { getAmpContainerID, ...restSelectors } = selectors;

	return {
		...store,
		actions: {
			...restActions,
			setAMPContainerID: setAmpContainerID,
		},
		selectors: {
			...restSelectors,
			getAMPContainerID: getAmpContainerID,
		},
	};
} )( baseModuleStore );

const store = Data.combineStores(
	baseModuleStore,
	accounts,
	containers,
	tags,
	settings,
	versions,
	createSnapshotStore( STORE_NAME ),
	service
);

// Register this store on the global registry.
Data.registerStore( STORE_NAME, store );

export const {
	INITIAL_STATE,
	actions,
	controls,
	reducer,
	resolvers,
	selectors,
} = store;

export default store;
