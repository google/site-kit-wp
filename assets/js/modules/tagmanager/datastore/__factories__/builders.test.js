/**
 * Factory builder tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import * as factories from '../../datastore/__factories__';
import { CONTEXT_AMP, CONTEXT_WEB } from '../constants';

describe( 'Google Tag Manager Builder fake data test', () => {
	it( 'should create the same data from factory builders on each test run', () => {
		// Set a hardcoded fingerprint value on each container to
		// prevent the factory from failing due to `Date.now()` changing
		// between test runs.
		function setStaticFingerprintOnFactories( containers ) {
			return containers.map( ( container ) => {
				return { ...container, fingerprint: '1676495375419' };
			} );
		}

		const account = factories.accountBuilder();
		const webContainers = factories.buildContainers(
			3,
			{ accountId: account.accountId, usageContext: [ CONTEXT_WEB ] } // eslint-disable-line sitekit/acronym-case
		);
		const ampContainers = factories.buildContainers(
			3,
			{ accountId: account.accountId, usageContext: [ CONTEXT_AMP ] } // eslint-disable-line sitekit/acronym-case
		);

		expect( {
			account,
			webContainers: setStaticFingerprintOnFactories( webContainers ),
			ampContainers: setStaticFingerprintOnFactories( ampContainers ),
		} ).toMatchSnapshot();
	} );
} );
