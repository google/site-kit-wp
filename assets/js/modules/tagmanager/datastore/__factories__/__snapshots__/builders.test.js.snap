// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Google Tag Manager Builder fake data test should create the same data from factory builders on each test run 1`] = `
{
  "account": {
    "accountId": "101",
    "name": "eligendi architecto veritatis",
    "path": "accounts/101",
  },
  "ampContainers": [
    {
      "accountId": "101",
      "containerId": "204",
      "fingerprint": "1676495375419",
      "name": "eaque ducimus eveniet",
      "path": "accounts/101/containers/204",
      "publicId": "GTM-FAKEBSD",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/204/workspaces?apiLink=container",
      "usageContext": [
        "amp",
      ],
    },
    {
      "accountId": "101",
      "containerId": "205",
      "fingerprint": "1676495375419",
      "name": "illo qui exercitationem",
      "path": "accounts/101/containers/205",
      "publicId": "GTM-FAKEW2G",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/205/workspaces?apiLink=container",
      "usageContext": [
        "amp",
      ],
    },
    {
      "accountId": "101",
      "containerId": "206",
      "fingerprint": "1676495375419",
      "name": "minus ad veritatis",
      "path": "accounts/101/containers/206",
      "publicId": "GTM-FAKEX9G",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/206/workspaces?apiLink=container",
      "usageContext": [
        "amp",
      ],
    },
  ],
  "webContainers": [
    {
      "accountId": "101",
      "containerId": "201",
      "fingerprint": "1676495375419",
      "name": "occaecati nam quae",
      "path": "accounts/101/containers/201",
      "publicId": "GTM-FAKEKZ3",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/201/workspaces?apiLink=container",
      "usageContext": [
        "web",
      ],
    },
    {
      "accountId": "101",
      "containerId": "202",
      "fingerprint": "1676495375419",
      "name": "iusto reprehenderit non",
      "path": "accounts/101/containers/202",
      "publicId": "GTM-FAKEHDB",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/202/workspaces?apiLink=container",
      "usageContext": [
        "web",
      ],
    },
    {
      "accountId": "101",
      "containerId": "203",
      "fingerprint": "1676495375419",
      "name": "voluptatem nisi nostrum",
      "path": "accounts/101/containers/203",
      "publicId": "GTM-FAKEDOU",
      "tagManagerUrl": "https://tagmanager.google.com/#/container/accounts/101/containers/203/workspaces?apiLink=container",
      "usageContext": [
        "web",
      ],
    },
  ],
}
`;
